core/src/ancestor_hashes_service.rs

```diff
@@ -0,0 +1,364 @@
+        outstanding_requests: &RwLock<OutstandingAncestorHashesRepairs>,
+        blockstore: &Blockstore,
+        duplicate_slots_reset_sender: &DuplicateSlotsResetSender,
+    ) {
+        // iter over the packets
+        packets.packets.iter().for_each(|packet| {
+            let from_addr = packet.meta.addr();
+            if let Ok(ancestor_hashes_response) =
+                limited_deserialize(&packet.data[..packet.meta.size - SIZE_OF_NONCE])
+            {
+                // Verify the response
+                let request_slot = repair_response::nonce(&packet.data[..packet.meta.size])
+                    .and_then(|nonce| {
+                        outstanding_requests.write().unwrap().register_response(
+                            nonce,
+                            &ancestor_hashes_response,
+                            timestamp(),
+                            // If the response is valid, return the slot the request
+                            // was for
+                            |ancestor_hashes_request| ancestor_hashes_request.0,
+                        )
+                    });
+
+                if request_slot.is_none() {
+                    stats.invalid_packets += 1;
+                    return;
+                }
+
+                // If was a valid response, there must be a valid `request_slot`
+                let request_slot = request_slot.unwrap();
+                stats.processed += 1;
+
+                // Check if we can make any decisions.
+                if let Occupied(mut ancestor_hashes_status_ref) =
+                    ancestor_hashes_request_statuses.entry(request_slot)
+                {
+                    if let Some(decision) = ancestor_hashes_status_ref.get_mut().add_response(
+                        &from_addr,
+                        ancestor_hashes_response.into_slot_hashes(),
+                        blockstore,
+                    ) {
+                        let potential_slots_to_dump = {
+                            // TODO: In the case of DuplicateAncestorDecision::ContinueSearch
+                            // This means all the ancestors were mismatched, which
+                            // means the earliest mismatched ancestor has yet to be found.
+                            //
+                            // In the best case scenario, this means after ReplayStage dumps
+                            // the earliest known ancestor `A` here, and then repairs `A`,
+                            // because we may still have the incorrect version of some ancestor
+                            // of `A`, we will mark `A` as dead and then continue the search
+                            // protocol through another round of ancestor repairs.
+                            //
+                            // However this process is a bit slow, so in an ideal world, the
+                            // protocol could be extended to keep searching by making
```

Comment:
what's stopping us from checking `ContinueSearch` here and sending another request?

Reply:
I think there's a race somewhere, but I can't think of it off the top of my head xD

Also more complexity for v1 to avoid.